lib/backend/memory/memory.go

```diff
@@ -290,25 +302,100 @@ func (m *Memory) DeleteRange(ctx context.Context, startKey, endKey backend.Key)
 	return nil
 }
 
-// GetRange returns query range
-func (m *Memory) GetRange(ctx context.Context, startKey, endKey backend.Key, limit int) (*backend.GetResult, error) {
-	if startKey.IsZero() {
-		return nil, trace.BadParameter("missing parameter startKey")
+func (m *Memory) Items(ctx context.Context, params backend.IterateParams) iter.Seq2[backend.Item, error] {
+	if params.StartKey.IsZero() {
+		err := trace.BadParameter("missing parameter startKey")
+		return func(yield func(backend.Item, error) bool) { yield(backend.Item{}, err) }
 	}
-	if endKey.IsZero() {
-		return nil, trace.BadParameter("missing parameter endKey")
+	if params.EndKey.IsZero() {
+		err := trace.BadParameter("missing parameter endKey")
+		return func(yield func(backend.Item, error) bool) { yield(backend.Item{}, err) }
 	}
+
+	limit := params.Limit
 	if limit <= 0 {
 		limit = backend.DefaultRangeLimit
 	}
-	m.Lock()
-	defer m.Unlock()
-	m.removeExpired()
-	re := m.getRange(ctx, startKey, endKey, limit)
-	if len(re.Items) == backend.DefaultRangeLimit {
+
+	const defaultPageSize = 1000
+	return func(yield func(backend.Item, error) bool) {
+		startKey := params.StartKey
+		if params.Descending {
+			startKey = params.EndKey
+		}
+
+		var totalCount, pageCount int
+		pageLimit := defaultPageSize
+		items := make([]backend.Item, pageLimit)
+		startItem := &btreeItem{Item: backend.Item{Key: startKey}}
+
+		treeIter := m.tree.AscendGreaterOrEqual
+		if params.Descending {
+			treeIter = m.tree.DescendLessOrEqual
+		}
+
+		for {
+			pageLimit = min(limit-totalCount, defaultPageSize)
+			pageCount = 0
+
+			m.Lock()
+			m.removeExpired()
+			var moreItems bool
+			treeIter(startItem, func(item *btreeItem) bool {
+				if params.Descending && item.Key.Compare(params.StartKey) < 0 {
+					startItem = item
+					return false
+				}
+
+				if !params.Descending && params.EndKey.Compare(item.Key) < 0 {
+					startItem = item
+					return false
+				}
+
+				if pageCount == pageLimit {
+					startItem = item
+					moreItems = true
+					return false
+				}
+
+				items[pageCount] = item.Item
+				pageCount++
+				return true
+			})
+			m.Unlock()
+
+			for _, item := range items[:pageCount] {
+				if !yield(item, nil) {
+					return
+				}
+
+				totalCount++
+				if limit != backend.NoLimit && totalCount >= limit {
+					return
+				}
```

Comment:
Either remove the `limit != backend.NoLimit` condition, or remove the logic earlier in this function that sets non-positive limit values to `backend.DefaultRangeLimit`, since they are contradictory.